### PR TITLE
[Messenger] Add options to `FailedMessagesShowCommand`

### PR DESCRIPTION
**Pull request by fguimier**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/14689

The goal to this PR is to improve the FailedMessagesShowCommand usability by adding two options:
- The `stats` option allows to display the number of messages by the class name and have a summary of all the failed messages.
* The `class-filter` option is a filter to only display the messages matching with the given class name.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 1]
```diff
@@ -98,12 +102,18 @@ private function listMessages(SymfonyStyle $io, int $max)
                 $errorMessage = $lastRedeliveryStampWithException->getExceptionMessage();
             }
 
-            $rows[] = [
-                $this->getMessageId($envelope),
-                \get_class($envelope->getMessage()),
-                null === $lastRedeliveryStamp ? '' : $lastRedeliveryStamp->getRedeliveredAt()->format('Y-m-d H:i:s'),
-                $errorMessage,
-            ];
+            $currentClassName = str_replace('\\', '', \get_class($envelope->getMessage()));
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: What's the purpose of this `str_replace`?

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 2]
```diff
@@ -98,12 +102,18 @@ private function listMessages(SymfonyStyle $io, int $max)
...
+
+            if (!$classFilter || $classFilter === $currentClassName) {
+                $rows[] = [
+                    $this->getMessageId($envelope),
+                    $currentClassName,
+                    null === $lastRedeliveryStamp ? '' : $lastRedeliveryStamp->getRedeliveredAt()->format(
+                        'Y-m-d H:i:s'
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: revert this change, keep a single line

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 3]
```diff
@@ -98,12 +102,18 @@ private function listMessages(SymfonyStyle $io, int $max)
...
 
-            $rows[] = [
-                $this->getMessageId($envelope),
-                \get_class($envelope->getMessage()),
-                null === $lastRedeliveryStamp ? '' : $lastRedeliveryStamp->getRedeliveredAt()->format('Y-m-d H:i:s'),
-                $errorMessage,
-            ];
+            $currentClassName = str_replace('\\', '', \get_class($envelope->getMessage()));
+
+            if (!$classFilter || $classFilter === $currentClassName) {
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: This check could have been done early: just after `foreach ($envelopes as $envelope) {`

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 4]
```diff
@@ -38,6 +38,8 @@ protected function configure(): void
             ->setDefinition([
                 new InputArgument('id', InputArgument::OPTIONAL, 'Specific message id to show'),
                 new InputOption('max', null, InputOption::VALUE_REQUIRED, 'Maximum number of messages to list', 50),
+                new InputOption('group', null, InputOption::VALUE_NONE, 'Group by message class'),
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: When looking at implementation, the wording is wrong. I expect the same behavior with messages grouped by class, but this looks like statistic per classes.

Maybe `stats` would be better?

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 5]
```diff
@@ -121,6 +131,38 @@ private function listMessages(SymfonyStyle $io, int $max)
         $io->comment('Run <comment>messenger:failed:show {id} -vv</comment> to see message details.');
     }
 
+    private function listMessagesPerClass(SymfonyStyle $io)
+    {
+        /** @var ListableReceiverInterface $receiver */
+        $receiver = $this->getReceiver();
+        $envelopes = $receiver->all();
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: Should take "max" into account or at least throw an exception if user call this command with both `max` and `group`

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 6]
```diff
@@ -121,6 +131,38 @@ private function listMessages(SymfonyStyle $io, int $max)
...
+
+        $countPerClass = [];
+
+        foreach ($envelopes as $envelope) {
+            if (!isset($countPerClass[\get_class($envelope->getMessage())])) {
+                $countPerClass[\get_class($envelope->getMessage())] = 0;
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: you could use `$countPerClass[$c = \get_class($envelope->getMessage())] = [$c, 0];` 
to avoid the `array_walk` below

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php (hunk 1 above)
Comment (fguimier) [reply src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: It allows to copy/paste quickly a fully qualified class name in the `class-filter` option like `--class-filter=App\Utility\MyClass`

But as the backslash are escaped from the option value, the direct comparison between `\get_class($envelope->getMessage()` and the  `class-filter` option value would have not work without this str_replace.

Comment (jderusse) [reply src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: sounds wrong to me to remove all the `\` in FQDN in case someone want copy/paste easily.

I suggest to allows filtering on part of the FQDN (strpos) instead

Comment (fguimier) [reply src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: Ok, I actually remove  this str_replace in my last commit to avoid any mistake

**Comment by fguimier** [comment]
@jderusse I've got an issue regarding the fabbot.io checking your CS.

After rebasing my branch with the latest version of 5.x to solve some conflicts, it detected several issues on FailedMessagesShowCommandTest but on 3 tests I did not update.
It would want to remove all the spaces in 3 heredoc blocks but I cannot do that as it would break these tests.

Is there solution to pass this test?

**Comment by OskarStark** [comment]
HI, it looks like your commiter email address is not associated with your Github Account:

CleanShot 2021-08-04 at 22 34 08@2x

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 7]
```diff
@@ -67,23 +69,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
         $receiver = $this->getReceiver();
         $envelopes = $receiver->all($max);
 
         $rows = [];
         foreach ($envelopes as $envelope) {
+            $currentClassName = \get_class($envelope->getMessage());
+
+            if ($classFilter && $classFilter !== $currentClassName) {
+                continue;
+            }
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php:114]: that's break pagination and the `max` options.
I think we should at least show how many messages have been ignored

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 8]
```diff
@@ -121,6 +131,31 @@ private function listMessages(SymfonyStyle $io, int $max)
...
+    {
+        /** @var ListableReceiverInterface $receiver */
+        $receiver = $this->getReceiver();
+        $envelopes = $receiver->all($max);
+
+        $countPerClass = [];
+
+        foreach ($envelopes as $envelope) {
+            if (!isset($countPerClass[\get_class($envelope->getMessage())])) {
+                $countPerClass[$c = \get_class($envelope->getMessage())] = [$c, 0];
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: `$c` could have been initialized above

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php [hunk 9]
```diff
@@ -121,6 +131,31 @@ private function listMessages(SymfonyStyle $io, int $max)
...
+            }
+
+            ++$countPerClass[\get_class($envelope->getMessage())][1];
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php]: `$c`

**Review by jderusse — COMMENTED** [review]
I like the idea ❤️

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php (hunk 7 above)
Comment (fguimier) [reply src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php:114]: I added more outputs including the count of displayed rows in case of the option class-filter is used and only if the max option is not reached.

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 10]
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 ---
 
  * Add `StopWorkerExceptionInterface` and its implementation `StopWorkerException` to stop the worker.
+* Add the options `stats` and `class-filter` to `FailedMessagesShowCommand`
```
Comment (derrabus) [inline src/Symfony/Component/Messenger/CHANGELOG.md]:
```suggestion
 * Add the options `stats` and `class-filter` to `FailedMessagesShowCommand`
```

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesShowCommandTest.php [hunk 11]
```diff
@@ -117,7 +117,7 @@ public function testBasicRunWithServiceLocator()
   Error         Things are bad!      
   Error Code    123                  
   Error Class   Exception            
-  Transport     async
+  Transport     async
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Tests/Command/FailedMessagesShowCommandTest.php:120]: The extra spaces should be removed

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesShowCommandTest.php [hunk 12]
```diff
@@ -198,7 +198,7 @@ public function testMultipleRedeliveryFailsWithServiceLocator()
...
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Tests/Command/FailedMessagesShowCommandTest.php:201]: The extra spaces should be removed

**Review by fabpot — CHANGES_REQUESTED** [review]
You should also revert the permission change on src/Symfony/Component/Messenger/Tests/Command/FailedMessagesShowCommandTest.php (get back to 0644)

**Comment by fabpot** [comment]
Closing in favor of #47008, where I've made some minor changes to finish it.
